lib/luaxtable.c

```diff
@@ -32,48 +14,190 @@
 #include <lauxlib.h>
 #include <lunatik.h>
 
+#include "luadata.h"
+#include "luarcu.h"
+#include "luaxtable.h"
+
 typedef enum luaxtable_type_e {
 	LUAXTABLE_TMATCH,
 	LUAXTABLE_TTARGET,
 } luaxtable_type_t;
 
 typedef struct luaxtable_s {
 	lunatik_object_t *runtime;
+	lunatik_object_t *skb;
 	union {
 		struct xt_match match;
 		struct xt_target target;
 	};
 	luaxtable_type_t type;
 } luaxtable_t;
 
-static bool luaxtable_match(const struct sk_buff *skb, struct xt_action_param *par)
+static struct {
+	lunatik_object_t *target;
+	lunatik_object_t *match;
+	bool match_fallback;
+	unsigned int target_fallback;
+} luaxtable_hooks = {NULL, NULL, false, XT_CONTINUE};
+
+static int luaxtable_invoke(lua_State *L, luaxtable_t *xtable, const char *op, int nargs, int nret)
+{
+	int base = lua_gettop(L) - nargs;
+
+	if (lunatik_getregistry(L, xtable) != LUA_TTABLE) {
+		pr_err("%s: could not find ops table\n", op);
+		goto err;
+	}
+
+	if (lua_getfield(L, -1, op) != LUA_TFUNCTION) {
+		pr_err("%s isn't defined\n", op);
+		goto err;
+	}
+
+	lua_insert(L, base + 1); /* op */
+	lua_pop(L, 1); /* table */
+
+	if (lua_pcall(L, nargs, nret, 0) != LUA_OK) {
+		pr_err("%s error: %s\n", op, lua_tostring(L, -1));
+		goto err;
+	}
+	return 0;
+err:
+	return -1;
+}
+
+static inline lunatik_object_t *luaxtable_getskb(lua_State *L, luaxtable_t *xtable)
 {
-	return true;
+	if (lunatik_getregistry(L, xtable->skb) != LUA_TUSERDATA)
+		return NULL;
+
+	return (lunatik_object_t *)lunatik_toobject(L, -1);
 }
 
-static int luaxtable_match_check(const struct xt_mtchk_param *par)
+static inline int luaxtable_pushskb(lua_State *L, luaxtable_t *xtable, const struct sk_buff *skb, uint8_t opt)
 {
+	lunatik_object_t *data = luaxtable_getskb(L, xtable);
+	if (data == NULL) {
+		pr_err("could not find skb\n");
+		return -1;
+	}
+
+	luadata_reset(data, skb->data, skb_headlen(skb), opt);
 	return 0;
 }
 
-static void luaxtable_match_destroy(const struct xt_mtdtor_param *par)
+static inline int luaxtable_pushparams(lua_State *L, const struct xt_action_param *par, luaxtable_t *xtable, const struct sk_buff *skb, uint8_t opt)
```

Comment:
this function has grow a bit.. I think there's no point in being an `inline` anymore.. btw, perhaps it's a good idea to move the body of `pushskb` inhere because it's sufficient small and it's not being used anywhere else.. it might clean up the code a bit.. what do you think?